runtime: support cgroup v2 metrics marshal guest metrics

yaoyinnan opened the pull request:
Support to use cgroup v2 metrics marshal guest metrics.

Fixes: #5998

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

liubin commented:
CI error:

> Error: metrics_test.go:57:13: cannot initialize 1 variables with 2 values (typecheck)
	metrics := statsToMetrics(&resp)

yaoyinnan commented:
/test

yaoyinnan commented:
> CI error:
> 
> > Error: metrics_test.go:57:13: cannot initialize 1 variables with 2 values (typecheck)
> > metrics := statsToMetrics(&resp)

I split `statsToMetrics()` and changed it to judge whether it is cgroup v1 or v2 in `marshalMetrics()`. The unit test `TestStatNetworkMetric()` is only tested when the cgroup is v1.
This modification is due to two problems: the first problem is that the stats returned by the agent do not contain `networkStats`, and there is no other way to obtain networkStats, refer to https://github.com/kata-containers/kata-containers /blob/main/src/agent/rustjail/src/container.rs#L856. The second problem is that the metrics of cgroup v2 do not include networkStats.

liubin commented:
/test-ubuntu-metrics
/test-vfio

bergwolf commented:
Please rebase this, thanks!

yaoyinnan commented:
> Please rebase this, thanks!

Rebased, thanks.

yaoyinnan commented:
/test

yaoyinnan commented:
/test

GabyCT commented:
/test-ubuntu

GabyCT commented:
/test-arm
/test-fc

luodw commented on src/runtime/pkg/containerd-shim-v2/metrics.go:
```diff
@@ -175,6 +258,28 @@ func copyBlkio(s []vc.BlkioStatEntry) []*cgroupsv1.BlkIOEntry {
 
+func copyBlkioV2(s []vc.BlkioStatEntry) []*cgroupsv2.IOEntry {
+	var ret []*cgroupsv2.IOEntry
+	var item *cgroupsv2.IOEntry
```
here item is null pointer?

yaoyinnan replied:
fixed

yaoyinnan commented:
/test

yaoyinnan commented:
/test

yaoyinnan commented:
/test-ubuntu

yaoyinnan commented:
/test-arm

yaoyinnan commented:
/test-ppc64le

yaoyinnan commented:
/test-power
